### PR TITLE
Enable user to turn off self registration, and enable admin user to add user.

### PR DESCRIPTION
**Pull request by wknet123**
added self-registration feature.

### CODE REVIEW

File: controllers/base.go [hunk 1]
```diff
@@ -96,12 +98,29 @@ func (b *BaseController) Prepare() {
...
 	if authMode == "" {
 		authMode = "db_auth"
 	}
 	b.Data["AuthMode"] = authMode
+
+	var selfRegistration = strings.ToLower(os.Getenv("SELF_REGISTRATION"))
+
+	switch selfRegistration {
+	case "on":
+		if sessionUserID != nil && sessionUserID.(int) != models.SYSADMIN {
```
Comment (reasonerjt) [inline controllers/base.go]: This is not needed as a bool variable's default value is false.

File: controllers/base.go [hunk 2]
```diff
@@ -96,12 +98,29 @@ func (b *BaseController) Prepare() {
...
+			enableAddUserByAdmin = false
+		} else {
+			enableAddUserByAdmin = true
+		}
+	case "":
```
Comment (reasonerjt) [inline controllers/base.go]: this is not needed

File: controllers/register.go [hunk 3]
```diff
@@ -33,16 +35,37 @@ type RegisterController struct {
...
 }
 
 // SignUp insert data into DB based on data in form.
 func (rc *CommonController) SignUp() {
+
+	if enableAddUserByAdmin {
+		sessionUserID := rc.GetSession("userId")
+		if sessionUserID == nil || sessionUserID.(int) != models.SYSADMIN {
+			log.Error("Self registration can only be used by admin user.\n")
+			rc.Redirect("/signIn", http.StatusFound)
```
Comment (reasonerjt) [inline controllers/register.go]: this is for ajax call and won't handle 302

File: controllers/base.go [hunk 4]
```diff
@@ -96,12 +98,29 @@ func (b *BaseController) Prepare() {
...
 		b.Data["Username"] = b.GetSession("username")
+		b.Data["UserId"] = sessionUserID.(int)
 	}
 	authMode := os.Getenv("AUTH_MODE")
 	if authMode == "" {
 		authMode = "db_auth"
 	}
 	b.Data["AuthMode"] = authMode
+
+	var selfRegistration = strings.ToLower(os.Getenv("SELF_REGISTRATION"))
```
Comment (reasonerjt) [inline controllers/base.go]: this has to be a global var so each package can read it.

**Comment by wknet123** [comment]
@xiahaoshawn please review my new updates, thanks.

File: controllers/base.go [hunk 5]
```diff
@@ -96,12 +98,27 @@ func (b *BaseController) Prepare() {
...
 	}
 	b.Data["AuthMode"] = authMode
+
+	strSelfRegistration := strings.ToLower(os.Getenv("SELF_REGISTRATION"))
+
+	if strSelfRegistration == "on" {
+		selfRegistration = true
+	}
+
+	if sessionUserID != nil && sessionUserID.(int) != adminUserID {
```
Comment (reasonerjt) [inline controllers/base.go]: b.Data["EnableAddUserByAdmin"] = !selfRegistration && isadmin

**Comment by xiahaoshawn** [comment]
which page should stay on after add user?

**Comment by wknet123** [comment]
Admin role user should always can add user when logged in whether the self-registration flag is on or off.

File: controllers/register.go [hunk 6]
```diff
@@ -34,22 +34,36 @@ type RegisterController struct {
...
-		pageTitleKey = "page_title_add_user"
+// AddUserController handles request for adding user with an admin role user
+type AddUserController struct {
+	BaseController
+}
+
+// Get renders the Sign In page, it only works if the auth mode is set to db_auth
+func (ac *AddUserController) Get() {
+
+	if !ac.BaseController.IsAdminLoginedUser {
```
Comment (reasonerjt) [inline controllers/register.go]: ac.IsAdminxxx should be fine

File: controllers/register.go [hunk 7]
```diff
@@ -61,8 +75,8 @@ func (rc *CommonController) SignUp() {
 		rc.CustomAbort(http.StatusForbidden, "")
```
Comment (reasonerjt) [inline controllers/register.go]: is it rc or ac???

File: controllers/register.go [hunk 8]
```diff
@@ -61,8 +75,8 @@ func (rc *CommonController) SignUp() {
...
 	}
 
-	if enableAddUserByAdmin && !isAdminLoginedUser {
-		log.Error("Self registration can only be used by admin user.\n")
+	if !(rc.BaseController.SelfRegistration || rc.BaseController.IsAdminLoginedUser) {
```
Comment (reasonerjt) [inline controllers/register.go]: I think it should be ac right?

File: controllers/base.go [hunk 9]
```diff
@@ -40,6 +40,8 @@ func (c *CommonController) Render() error {
 type BaseController struct {
 	beego.Controller
 	i18n.Locale
+	SelfRegistration   bool
+	IsAdminLoginedUser bool
```
Comment (reasonerjt) [inline controllers/base.go]: I think this var is too long, IsAdmin should convey the meaning...

File: controllers/register.go [hunk 10]
```diff
@@ -34,22 +34,36 @@ type RegisterController struct {
 // Get renders the Sign In page, it only works if the auth mode is set to db_auth
 func (rc *RegisterController) Get() {
 
-	if enableAddUserByAdmin && !isAdminLoginedUser {
-		log.Error("Self registration can only be used by admin user.\n")
+	if !rc.BaseController.SelfRegistration {
```
Comment (reasonerjt) [inline controllers/register.go]: don't need the BaseController

File: controllers/register.go [hunk 11]
```diff
@@ -32,16 +33,53 @@ type RegisterController struct {
 
 // Get renders the Sign In page, it only works if the auth mode is set to db_auth
 func (rc *RegisterController) Get() {
+
+	if !rc.BaseController.SelfRegistration {
+		log.Error("Registration can only be used by admin user when self-registrion is off.\n")
```
Comment (reasonerjt) [inline controllers/register.go]: Error message is wrong

File: controllers/register.go [hunk 12]
```diff
@@ -32,16 +33,53 @@ type RegisterController struct {
...
 
 // SignUp insert data into DB based on data in form.
 func (rc *CommonController) SignUp() {
+
+	authMode := os.Getenv("AUTH_MODE")
+	if !(authMode == "" || authMode == "db_auth") {
+		rc.CustomAbort(http.StatusForbidden, "")
+	}
+
+	if !(rc.BaseController.SelfRegistration || rc.BaseController.IsAdminLoginedUser) {
```
Comment (reasonerjt) [inline controllers/register.go]: BaseController is not needed?

File: controllers/base.go [hunk 13]
```diff
@@ -96,12 +100,33 @@ func (b *BaseController) Prepare() {
...
 	}
 	b.Data["AuthMode"] = authMode
+
+	selfRegistration := strings.ToLower(os.Getenv("SELF_REGISTRATION"))
+
+	if selfRegistration == "on" {
+		b.SelfRegistration = true
+	}
+
+	if sessionUserID != nil {
```
Comment (reasonerjt) [inline controllers/base.go]: you should be two chunks of if sessionUserID != nil  into one

File: controllers/base.go [hunk 14]
```diff
@@ -96,12 +100,33 @@ func (b *BaseController) Prepare() {
...
+		var err error
+		b.IsAdminLoginedUser, err = dao.IsAdminRole(sessionUserID)
```
Comment (reasonerjt) [inline controllers/base.go]: cast it to int?

File: controllers/register.go [hunk 15]
```diff
@@ -32,16 +33,53 @@ type RegisterController struct {
...
+}
+
+// Get renders the Sign In page, it only works if the auth mode is set to db_auth
+func (ac *AddUserController) Get() {
+
+	if !ac.BaseController.IsAdminLoginedUser {
+		ac.Redirect("/signIn", http.StatusFound)
+	}
+
+	authMode := os.Getenv("AUTH_MODE")
```
Comment (reasonerjt) [inline controllers/register.go]: Move authMode to basic Controller so there won't be dup code?

File: controllers/register.go [hunk 16]
```diff
@@ -32,16 +33,53 @@ type RegisterController struct {
...
 func (rc *CommonController) SignUp() {
+
+	authMode := os.Getenv("AUTH_MODE")
+	if !(authMode == "" || authMode == "db_auth") {
+		rc.CustomAbort(http.StatusForbidden, "")
+	}
+
+	if !(rc.BaseController.SelfRegistration || rc.BaseController.IsAdminLoginedUser) {
+		log.Error("Registration can only be used by admin role user when self-registration is off.\n")
+		rc.CustomAbort(http.StatusForbidden, "")
```
Comment (reasonerjt) [inline controllers/register.go]: why is CommonController's instance called rc?

File: controllers/base.go [hunk 17]
```diff
@@ -38,6 +40,9 @@ func (c *CommonController) Render() error {
 type BaseController struct {
 	beego.Controller
 	i18n.Locale
+	SelfRegistration bool
```
Comment (ywk253100) [inline controllers/base.go]: Can we change "SelfRegistration" and "AuthMode" to global variables? We will read them from env only once in this way.

File: controllers/register.go [hunk 18]
```diff
@@ -32,35 +32,70 @@ type RegisterController struct {
 
 // Get renders the Sign In page, it only works if the auth mode is set to db_auth
 func (rc *RegisterController) Get() {
-	authMode := os.Getenv("AUTH_MODE")
-	if authMode == "" || authMode == "db_auth" {
+
+	if !rc.SelfRegistration {
+		log.Error("Registration is disabled when self-registration is off.\n")
```
Comment (ywk253100) [inline controllers/register.go]: log.Error is not proper here, maybe Info or Debug is more suitable.

File: controllers/register.go [hunk 19]
```diff
@@ -32,35 +32,70 @@ type RegisterController struct {
...
+		rc.Redirect("/signIn", http.StatusFound)
```
Comment (ywk253100) [inline controllers/register.go]: We should response a 403 and redirect a 403 page, or users will confused.

File: controllers/register.go [hunk 20]
```diff
@@ -32,35 +32,70 @@ type RegisterController struct {
...
+	}
+}
+
+// AddUserController handles request for adding user with an admin role user
+type AddUserController struct {
+	BaseController
+}
+
+// Get renders the Sign In page, it only works if the auth mode is set to db_auth
+func (ac *AddUserController) Get() {
```
Comment (ywk253100) [inline controllers/register.go]: AddUserController is not needed. RegisterController is enough.

**Pull request merged**